File: packages/spreadsheet-engine/src/archive.ts
import {
  BlobWriter,
  Uint8ArrayReader,
  Uint8ArrayWriter,
  ZipReader,
  ZipWriter,
} from "@zip.js/zip.js";
import { defaultDocumentLimits } from "@baseblocks/anydoc-contracts";

export type SpreadsheetOpenLimits = Readonly<{
  maxEntries?: number;
  maxInputBytes?: number;
  maxPartBytes?: number;
  maxUncompressedBytes?: number;
  maxCells?: number;
}>;

const DEFAULT_LIMITS = {
  maxEntries: 10_000,
  maxInputBytes: 100 * 1024 * 1024,
  maxPartBytes: defaultDocumentLimits.archive.maxPartBytes,
  maxUncompressedBytes: defaultDocumentLimits.archive.maxUncompressedBytes,
  maxCells: defaultDocumentLimits.maxSpreadsheetCells,
} as const;

function unsafePath(name: string): boolean {
  return name.startsWith("/") || name.startsWith("\\") || name.split(/[\\/]/u).includes("..");
}

export class OoxmlArchive {
  readonly #parts: Map<string, Uint8Array>;

  private constructor(parts: Map<string, Uint8Array>) {
    this.#parts = parts;
  }

  static async open(bytes: Uint8Array, limits: SpreadsheetOpenLimits = {}): Promise<OoxmlArchive> {
    const resolved = { ...DEFAULT_LIMITS, ...limits };
    if (bytes.byteLength === 0) throw new Error("Workbook is empty.");
    if (bytes.byteLength > resolved.maxInputBytes)
      throw new Error("Workbook exceeds the compressed size limit.");
    const reader = new ZipReader(new Uint8ArrayReader(bytes));
    try {
      const entries = await reader.getEntries();
      if (entries.length > resolved.maxEntries)
        throw new Error("Workbook contains too many ZIP entries.");
      let total = 0;
      const parts = new Map<string, Uint8Array>();
      for (const entry of entries) {
        if (entry.directory) continue;
        if (unsafePath(entry.filename))
          throw new Error(`Workbook contains an unsafe ZIP path: ${entry.filename}`);
        if (entry.encrypted) throw new Error("Encrypted workbooks are not supported.");
        if (parts.has(entry.filename))
          throw new Error(`Workbook contains a duplicate ZIP entry: ${entry.filename}`);
        if (entry.uncompressedSize > resolved.maxPartBytes)
          throw new Error(`Workbook part exceeds the size limit: ${entry.filename}`);
        total += entry.uncompressedSize;
        if (total > resolved.maxUncompressedBytes)
          throw new Error("Workbook exceeds the expanded size limit.");
        const partWriter = new Uint8ArrayWriter();
        (partWriter as Uint8ArrayWriter & { maxSize: number }).maxSize = resolved.maxPartBytes;
        const data = await entry.getData?.(partWriter);
        if (!data) throw new Error(`Unable to read workbook part: ${entry.filename}`);
        if (data.byteLength > resolved.maxPartBytes)
          throw new Error(`Workbook part exceeds the size limit: ${entry.filename}`);
        total += data.byteLength - entry.uncompressedSize;
        if (total > resolved.maxUncompressedBytes)
          throw new Error("Workbook exceeds the expanded size limit.");
        parts.set(entry.filename, data);
      }
      return new OoxmlArchive(parts);
    } finally {
      await reader.close();
    }
  }

  cloneParts(): Map<string, Uint8Array> {
    return new Map(this.#parts);
  }

  has(name: string): boolean {
    return this.#parts.has(name);
  }

  names(): readonly string[] {
    return [...this.#parts.keys()].sort();
  }

  part(name: string): Uint8Array {
    const value = this.#parts.get(name);
    if (!value) throw new Error(`Workbook part is missing: ${name}`);
    return value;
  }

  text(name: string): string {
    return new TextDecoder().decode(this.part(name));
  }

  async export(
    overrides: ReadonlyMap<string, Uint8Array>,
    removals: ReadonlySet<string> = new Set(),
  ): Promise<Uint8Array> {
    const writer = new ZipWriter(
      new BlobWriter("application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"),
    );
    const names = new Set([...this.#parts.keys(), ...overrides.keys()]);
    for (const name of [...names].sort()) {
      if (removals.has(name)) continue;
      const data = overrides.get(name) ?? this.#parts.get(name);
      if (data) await writer.add(name, new Uint8ArrayReader(data), { level: 6 });
    }
    const blob = await writer.close();
    return new Uint8Array(await blob.arrayBuffer());
  }
}

File: spreadsheet-view/src/archive.rs
//! ZIP archive reading under the engine's resource limits.

use std::collections::HashSet;
use std::io::Read;

/// A fully expanded, limit-checked workbook archive.
pub struct OoxmlArchive {
    parts: std::collections::BTreeMap<String, Vec<u8>>,
}

fn unsafe_path(name: &str) -> bool {
    name.starts_with('/')
        || name.starts_with('\\')
        || name.split(['\\', '/']).any(|part| part == "..")
}

impl OoxmlArchive {
    pub fn open(bytes: &[u8], limits: &crate::limits::ResolvedLimits) -> Result<Self, String> {
        if bytes.is_empty() {
            return Err("Workbook is empty.".to_string());
        }
        if bytes.len() as u64 > limits.max_input_bytes {
            return Err("Workbook exceeds the compressed size limit.".to_string());
        }
        let reader = std::io::Cursor::new(bytes);
        let mut zip = zip::ZipArchive::new(reader)
            .map_err(|error| format!("Workbook could not be read: {error}"))?;
        if zip.len() as u64 > limits.max_entries {
            return Err("Workbook contains too many ZIP entries.".to_string());
        }
        let mut total: u64 = 0;
        let mut seen = HashSet::new();
        let mut parts = std::collections::BTreeMap::new();
        for index in 0..zip.len() {
            let mut entry = zip.by_index(index).map_err(|error| error.to_string())?;
            let name = entry.name().to_string();
            if entry.is_dir() {
                continue;
            }
            if unsafe_path(&name) {
                return Err(format!("Workbook contains an unsafe ZIP path: {name}"));
            }
            if entry.encrypted() {
                return Err("Encrypted workbooks are not supported.".to_string());
            }
            if !seen.insert(name.clone()) {
                return Err(format!("Workbook contains a duplicate ZIP entry: {name}"));
            }
            if entry.size() > limits.max_part_bytes {
                return Err(format!("Workbook part exceeds the size limit: {name}"));
            }
            total += entry.size();
            if total > limits.max_uncompressed_bytes {
                return Err("Workbook exceeds the expanded size limit.".to_string());
            }
            let mut data = Vec::with_capacity(entry.size() as usize);
            entry.read_to_end(&mut data).map_err(|error| error.to_string())?;
            if data.len() as u64 > limits.max_part_bytes {
                return Err(format!("Workbook part exceeds the size limit: {name}"));
            }
            total += data.len() as u64 - entry.size();
            if total > limits.max_uncompressed_bytes {
                return Err("Workbook exceeds the expanded size limit.".to_string());
            }
            parts.insert(name, data);
        }
        Ok(Self { parts })
    }

    pub fn has(&self, name: &str) -> bool {
        self.parts.contains_key(name)
    }

    /// Sorted part names, mirroring `archive.names()`.
    pub fn names(&self) -> Vec<&str> {
        self.parts.keys().map(String::as_str).collect()
    }

    pub fn text(&self, name: &str) -> Result<String, String> {
        let data = self.part(name)?;
        Ok(String::from_utf8_lossy(data).into_owned())
    }

    pub fn part(&self, name: &str) -> Result<&Vec<u8>, String> {
        self.parts.get(name).ok_or_else(|| format!("Workbook part is missing: {name}"))
    }
}
